### ISSUE TITLE
DeepChem extensibility issues

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#1076]
I've heard from multiple teams that while they've found DeepChem very useful, they've often had to significantly modify the library in order to make use of it. This process usually takes them far enough from DeepChem's core structure that it's not even clear how they could merge their changes back into the main library without breaking large amounts of code.

It's worth asking why multiple users have run into this issue. I've proposed some changes in #1075 that I think will help a little. I think #1038 will help as well, but it looks like it's going to take a good chunk of development effort to make happen.

If you have thoughts on how we can improve the API to make it more extensible and useful, please chime in on this thread.

CC @evanfeinberg

**Comment by peastman** [comment]
Can you provide some concrete examples of things people couldn't do without modifying it?

**Comment by rbharath** [comment]
Have contacted one of the teams to ask them to comment here if willing.

@evanfeinberg Would love your thoughts on this thread as well.

**Comment by ankitvgupta** [comment]
@rbharath Thanks so much for the great library - we are huge fans! 

We are building some production systems that involve using tf.Estimator, along with a few tf.Keras-based models for one-off cases here and there. As a result, incorporating TensorGraph layers seemed like reinventing the wheel a little bit. I think that your proposal for having more interoperability between TensorGraph and Keras, or moving to Keras entirely would address this challenge.

That said, we have found the models implemented in deepchem to be super useful, and very easy to port to tf.Estimators if needed. We are looking into making contributions into deepchem directly as well, and will definitely circle back about that in the coming weeks. 

Thanks again for all your efforts!

**Comment by rbharath** [comment]
@ankitvgupta Thanks for the detailed feedback! I'd love to hear your thoughts on tf.Estimator vs. tf.Keras. Are tf.Keras models interoperable with tf.Estimator? Do you think there would be gain from making DeepChem interoperate well with tf.Estimator?

**Comment by ankitvgupta** [comment]
@rbharath Currently, the story around the intersection of tf.Keras and tf.Estimator is a little muddled. Estimator is essentially a wrapper around the vanilla Tensorflow which allows for convenient experimentation, as well as distributed multi-node training. It also integrates very closely with the tf.data.Dataset API, which has given us various performance benefits.

tf.Keras supposedly works pretty well with Estimator, since the tf.keras library ships with a method to convert Keras models to Estimators (`tf.keras.estimator.model_to_estimator`), though I haven't tried that personally. That said, I believe that since you can use individual keras layers on tensorflow tensors, it'll work just fine in an Estimator, since an Estimator can wrap any valid tensorflow graph.

The main reason we didn't want to depend too highly on DeepChem's TensorGraph is so that we could use as much of the automated experimentation tools that are now shipping with Tensorflow directly. For example, `tf.contrib.learn.Experiment` can be used to further wrap an Estimator and manage all of the flow around training/validation/inference/logging, which has simplified our codebase quite a bit.

I do believe that Tensorflow is moving pretty strongly in the direction of making tf.Estimator central, especially since I see new tutorials written in that way. In my opinion, the best way to create interoperability is to have a really good separation of model functions. For example, if a model function can take input features and a dictionary of hyperparameters and return the final layer logits using any sequence of Tensorflow graph operations, that should be interoperable with any of the wrapper techniques like Estimator.

**Comment by peastman** [comment]
@ankitvgupta One possibility I'm looking into right now is adding a `TensorGraph.make_estimator()` method that would create an Estimator from any model, similar to how the Keras `model_to_estimator()` function works.  Does that sound like a good approach to you, or do you think something else would be better?  That would address the use case of integrating with tools designed to work with estimators.  But it wouldn't be able to support a lot of TensorGraph's more advanced capabilities like submodels, serialization, advanced support for recurrent layers, etc.

**Comment by ankitvgupta** [comment]
@peastman That seems like a reasonable starting point. The fundamental question that I think is more important is whether building that functionality is more useful in the long-term than building the Keras integration that @rbharath has discussed elsewhere. Intuitively speaking, I am generally a bit concerned about maintaining TensorGraph over the long term when packages like Keras exist and can do much of its functionality, especially with a large team of engineers maintaining them.

My guess is deepchem can be most useful if the contributors invest most of their time in what makes it unique: the chemistry, rather than dealing with all of the intricacies of writing/maintaining a Tensorflow wrapper long-term. That said, if you think that writing a `TensorGraph.make_estimator()` is relatively straightforward, I think that would be a nice stopgap solution to allow people to use new Tensorflow tooling while being able to keep their old models. I would be happy to provide input in making sure that the estimator addresses many of the intricate use-cases that are necessary, such as having a `model_fn` that can take in multiple feature matrices (as is needed in many models in deepchem).

**Comment by peastman** [comment]
I would love to be able to switch over to Keras.  Unfortunately, my investigations have run into some serious roadblocks (see https://github.com/deepchem/deepchem/issues/1038).  For example, Keras has built in limitations that make it incapable of handling multitask networks without hacks that would lead to a really ugly API.  It also is incompatible with the metrics in DeepChem (most of which are just imported from sklearn), but only provides a tiny handful of its own metrics.  More important, all my attempts to raise these issues with the Keras developers (both on the discussion forum and the github repository) have been totally ignored.  As have similar feature requests from other people raising the same issues.  I'm getting really bad vibes about the Keras developer community.  I don't think we can rely on them.

So for the moment, I think we're forced to stick with TensorGraph.  Given that, improving its ability to interact with other Tensorflow based tools seems like a good idea.

**Comment by peastman** [comment]
I've posted a preliminary implementation of this API at #1106.

**Comment by rbharath** [comment]
Closing this old discussion since a lot of these changes have been made already.
